### PR TITLE
Add multiple users support for prototyping in Heroku

### PR DESCRIPTION
**Pull request by pmanrubia**
Adds multiple users support in Heroku deployments

### Background

The support for login is provided by [GDS-SSO][1] which is a gem that provides everything needed to integrate with [Signon][2]

In #384, we created a `rake` task to:

1. Deploy a branch to `Heroku`
2. Populate `Heroku` with a subset of the `Content::Item`s and `Link`s (about 1000)

Once a branch is deployed to `Heroku` it runs in `development` mode, so [GDS::SSO][1] mocks [Signon][2] and [authenticates automatically to the first user of the database][3]. This means, that only one user will exist in Heroku, and we have no easy way to authenticate other users.

This PR overcomes this issue by providing a simple authentication mechanism based on an URL parameter: `heroku_user_id` and the user session. This is obviously intended to be used only in `Heroku`.

### About the solution

1. I'd rather not creating a PR to support multiple users in [GDS-SSO][1] as this would be a temporary solution that would need to change once we move to PAAS.
2. I have created a helper class `lib/heroku.rb` that detects when the code runs in Heroku: checks for the existence of the environment variable `RUNNING_IN_HEROKU` and that we are not in `production` mode.
3. I have added an initialiser `config/initializers/heroku.rb` that overrides the default security mechanisms that are in place. 
4. I've removed the login information from the admin template, and I have added support for a select combobox to choose the user.

### After

image

[1]: https://github.com/alphagov/gds-sso
[2]: https://github.com/alphagov/signon
[3]: https://github.com/alphagov/gds-sso/blob/82ed8e549739a5ac807565ac49123e5dffff158d/lib/gds-sso/warden_config.rb#L83

### CODE REVIEW

File: lib/heroku.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Heroku
+  def self.create_users(organisation_id)
+    organisation = Content::Item.find(organisation_id)
+
+    User.create!(
+      name: 'AT user',
+      email: 'uat@gds.com',
```
Comment (chao-xian) [inline lib/heroku.rb]: `.com`??? `.gov.uk` surely 😉

File: lib/heroku.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+      uid: SecureRandom.uuid,
```
Comment (mgrassotti) [inline lib/heroku.rb]: Wouldn't this eventually fail in the (very rare) situation in that the same uuid is generated twice?
We could add a loop to check if a uuid was already generated. I mean something like:
```
def generate_uid
  loop do
     random_uid = SecureRandom.urlsafe_base64(nil, false)
     break random_uid unless User.exists?(uid: random_uid)
   end
 end
```

File: spec/lib/heroku_spec.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    context 'in development' do
+
+      it 'returns true if `RUNNING_IN_HEROKU` is present' do
+        expect(ENV).to receive(:[]).with("RUNNING_IN_HEROKU").and_return("TRUE")
+
+        expect(described_class.enabled?).to be true
+      end
+      it 'returns false otherwise' do
+        expect(ENV).to receive(:[]).with("RUNNING_IN_HEROKU").and_return(nil)
```
Comment (chao-xian) [inline spec/lib/heroku_spec.rb]: Is this supposed to have `.with("RUNNING_IN_HEROKU")`?

File: config/initializers/heroku.rb [hunk 4]
```diff
@@ -1,24 +1,28 @@
 if Heroku.enabled?
   ApplicationController.class_eval do
-    before_action :ensure_environment
+    before_action :force_development!
```
Comment (chao-xian) [inline config/initializers/heroku.rb]: Should this be a separate commit from the disabling of the GA?

File: lib/tasks/heroku.rake [hunk 5]
```diff
@@ -43,12 +43,12 @@ namespace :heroku do
     system %{heroku pg:psql -c "DELETE FROM CONTENT_ITEMS" --app #{app_name}}
     system %{heroku pg:psql -c "DELETE FROM USERS" --app #{app_name}}
 
-    # Creates default user
-    system %{heroku run rails runner "Heroku.create_users(#{organisation.id})" --app #{app_name}}
-
     #Import Contetn Items and Links
```
Comment (chao-xian) [inline lib/tasks/heroku.rake]: Small typo

File: spec/lib/heroku_spec.rb (hunk 3 above)
Comment (pmanrubia) [reply spec/lib/heroku_spec.rb]: I would say so. If the environment variable is not set then it always should return false?

File: lib/heroku.rb (hunk 2 above)
Comment (pmanrubia) [reply lib/heroku.rb]: @mgrassotti ok, I will try with the following then:

```ruby
uid: "uid-#{index}"
```

**Comment by pmanrubia** [comment]
@mgrassotti, @chao-xian thank you for reviewing it 👍

File: spec/lib/heroku_spec.rb (hunk 3 above)
Comment (chao-xian) [reply spec/lib/heroku_spec.rb]: I was confused as I didn't know that `expect` can also implicitly behave like an `allow` as well as assert. Thanks for clarifying!
